consensus/polybft/proposer_calculator_test.go

```diff
@@ -0,0 +1,510 @@
+
+	proposerAddressR2, err := pc.CalcProposer(2, 0)
+	require.NoError(t, err)
+	assert.Equal(t, metadata[2].Address, proposerAddressR2)
+
+	proposerAddressR3, err := pc.CalcProposer(3, 0)
+	require.NoError(t, err)
+	assert.Equal(t, metadata[1].Address, proposerAddressR3)
+
+	proposerAddressR4, err := pc.CalcProposer(4, 0)
+	require.NoError(t, err)
+	assert.Equal(t, metadata[4].Address, proposerAddressR4)
+
+	proposerAddressR5, err := pc.CalcProposer(5, 0)
+	require.NoError(t, err)
+	assert.Equal(t, metadata[3].Address, proposerAddressR5)
+
+	proposerAddressR6, err := pc.CalcProposer(6, 0)
+	require.NoError(t, err)
+	assert.Equal(t, metadata[0].Address, proposerAddressR6)
+}
+
+func TestProposerCalculator_SamePriority(t *testing.T) {
+	t.Parallel()
+
+	keys, err := bls.CreateRandomBlsKeys(5)
+	require.NoError(t, err)
+
+	// at some point priorities will be the same and bytes address will be compared
+	vs, err := NewValidatorSet([]*ValidatorMetadata{
+		{
+			BlsKey:      keys[0].PublicKey(),
+			Address:     types.Address{0x1},
+			VotingPower: 1,
+		},
+		{
+			BlsKey:      keys[1].PublicKey(),
+			Address:     types.Address{0x2},
+			VotingPower: 2,
+		},
+		{
+			BlsKey:      keys[2].PublicKey(),
+			Address:     types.Address{0x3},
+			VotingPower: 3,
+		},
+	}, hclog.NewNullLogger())
+	require.NoError(t, err)
+
+	snapshot := NewProposerSnapshot(0, vs.Accounts())
+	pc := NewProposerCalculatorFromSnapshot(snapshot, hclog.NewNullLogger())
+
+	proposerR0, err := pc.CalcProposer(0, 0)
+	require.NoError(t, err)
+	assert.Equal(t, types.Address{0x3}, proposerR0)
+
+	proposerR1, err := pc.CalcProposer(1, 0)
+	require.NoError(t, err)
+	assert.Equal(t, types.Address{0x2}, proposerR1)
+
+	proposerR2, err := pc.CalcProposer(2, 0)
+	require.NoError(t, err)
+	assert.Equal(t, types.Address{0x1}, proposerR2)
+
+	proposerR2, err = pc.CalcProposer(2, 0) // call again same round
+	require.NoError(t, err)
+	assert.Equal(t, types.Address{0x1}, proposerR2)
+}
+
+func TestProposerCalculator_ProposerSelection1(t *testing.T) {
+	t.Parallel()
+
+	const numberOfIteration = 99
+
+	keys, err := bls.CreateRandomBlsKeys(3)
+	require.NoError(t, err)
+
+	vset, err := NewValidatorSet([]*ValidatorMetadata{
+		{
+			BlsKey:      keys[0].PublicKey(),
+			Address:     types.Address{0x1},
+			VotingPower: 1000,
+		},
+		{
+			BlsKey:      keys[1].PublicKey(),
+			Address:     types.Address{0x2},
+			VotingPower: 300,
+		},
+		{
+			BlsKey:      keys[2].PublicKey(),
+			Address:     types.Address{0x3},
+			VotingPower: 330,
+		},
+	}, hclog.NewNullLogger())
+	require.NoError(t, err)
+
+	snapshot := NewProposerSnapshot(4, vset.Accounts())
+	pc := NewProposerCalculatorFromSnapshot(snapshot, hclog.NewNullLogger())
+
+	var proposers = make([]types.Address, numberOfIteration)
+
+	for i := uint64(0); i < numberOfIteration; i++ {
+		proposers[i], err = pc.CalcProposer(i, 4)
+		require.NoError(t, err)
+	}
+
+	expected := []types.Address{
```

Comment:
What does this expected array represent?

Reply:
Those are addresses of proposers which are expected to be calculated in that UT. @0xSasaPrsic to confirm.

Reply:
Yes, expected addresses in the order in which it is expected for the proposer to be selected.